Check current status of split view before toggling

giordano opened the pull request:
Fix #353

giordano commented on python/nammu/view/AtfAreaView.py:
```diff
@@ -129,50 +129,78 @@ def toggle_split_arabic(self, split_orientation,
+        self.vert_scroll = self.container.getVerticalScrollBar()
+        listener = atfAreaAdjustmentListener(self)
+        self.vert_scroll.addAdjustmentListener(listener)
+        menu = self.controller.controller.menuController.view.getMenu(3)
```
I'm not particularly happy with hard-coding the position of the menu and the items (lines below).  However, for the time being I couldn't come up with a solution different from "iterate over all menus/items and find the one matching this hard-coded text" which has the same flaw.

ageorgou replied:
This doesn't fully solve the problem, but you could hide some of this in the `MenuController` (through methods like `disable_xxxx_split`). At least then it's closer to the source, and you could define the positions as constants in there.

ageorgou replied:
Actually, here is a better solution:

- In `MenuView`, keep track of the position of each menu as they are created. For example, have a dictionary `menu_positions` that maps menu names to indices, and populate it during the loop in the [constructor loop](https://github.com/oracc/nammu/blob/1ae222ca93e5f35ff4f159dd75aa7e04ebc31915/python/nammu/view/MenuView.py#L111-L117).
- Add a method to `MenuView` that gets a menu by its name. This could just be
```python
def getMenuByName(self, name):
    return self.menu_positions[name]
```
- Do something similar in each `Menu`. For example, in the constructor:
```python
self.menu_positions = {name: index for (index, name) in enumerate(menuItems)}
```
(there's a chance this doesn't work in Python 2.7, but a loop would)

- In the `MenuView` again, add a method to en/disable a particular option in a menu:
```python
def disable_item(self, menu_name, menu_item):
    menu_index = self.menu_positions[menu_name]
    menu_item_index = self.getMenu(menu_index).menu_positions[menu_item]
    self.getMenu(menu_index).getItem(menu_item_index).setEnabled(False)
```

- Then, in the `MenuController`, you can have a method like:
```python
def disable_split_options(self, horizontal=False, vertical=False, arabic=False):
    # choose the right item names here
    for option in options:
        self.view.disable_item("Window", option)
```

giordano commented:
I added the improvements suggested by Anastasis (slightly changed, but the bulk was good, thanks a lot!), I only need to add some tests :see_no_evil:

giordano commented:
I think this is good for review now.  I added a few tests, that are eventually passing on Travis (after several restarts :roll_eyes:)

raquel-ucl commented on python/nammu/controller/MenuController.py:
```diff
@@ -34,6 +34,18 @@ def __init__(self, mainController):
+    def enable_split_options(self, horizontal=True,
+                             vertical=True, arabic=True):
+        """
+        Set visibility of split menu items.
```
Maybe replace with "Show split menu items as enabled or disabled" or something like that so it's more clear.

raquel-ucl commented on python/nammu/view/MenuView.py:
```diff
@@ -107,15 +107,38 @@ def __init__(self, controller):
     def __getattr__(self, name):
         return getattr(self.controller, name)
+
+    def getMenuByName(self, name):
```
Try to not use camel case for variable names and methods to follow python convention. We should only use those for Java method calls.

raquel-ucl commented on python/nammu/test/test_nammu.py:
```diff
@@ -412,3 +412,81 @@ def test_redo_arabic_pane(self, arabic, nammu):
+        assert view.container.getOrientation() == JSplitPane.VERTICAL_SPLIT
+        assert not toggleVertical.isEnabled()
+        assert not toggleHorizontal.isEnabled()
+        assert not toggleArabic.isEnabled()
```
TODO: Add test that when there is an arabic file and toggling vert/horiz won't make disappear the arabic pane.

raquel-ucl commented:
Also we can add a separator in the Window menu between the hide/show options and the toggling options.

raquel-ucl commented:
Found a small logic inconsistency:

1. Open Nammu
1. Choose Toggle Arabic pane
1. Open arabic.atf
1. The arabic pane and the arabic part of the file are not displayed 

If you try to open the arabic.atf file again, it will work fine. It just doesn't seem to work when opening from arabic toggle mode.

giordano commented:
I think I addressed all comments.

Regarding the last one, my solution has been to add a `force` argument to `NammuController.arabic` and call `arabic(force=True)` when opening an Arabic file, see https://github.com/oracc/nammu/pull/384/commits/43aa2caec405c81a38aaa3da783358236ee42d21#diff-2bd150ec3e0b82e145346223be946f73.  I couldn't find any reliable control for fixing this.

I find the whole toggling mechanism  a bit perverse as it involves non-mutual-exclusive features to be toggled.  Now there are functions to just do the thing (like `atfAreaView.setup_edit_area_split` and `atfAreaView.setup_edit_area_no_split`), and there are some checks before actually toggling, instead of blindly toggling whatever it was before.
